Customized warnings in cmake so XRT compiles with clang.

HFTrader opened the pull request:
Signed-off-by: Henrique Bucher <henry@vitorian.com>

<!-- Thanks for sending a pull request! Please fill out below, remove sections that don't apply for your pull request.  -->
#### Problem solved by the commit

XRT now compiles with clang

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered

This solves issue #6278 
https://github.com/Xilinx/XRT/issues/6278

#### How problem was solved, alternative solutions (if any) and why they were rejected

Clang issues many more warnings than Gcc. As you're compiling with -Werror it breaks the build. The solution was to detect that when the compiler is Clang, it adds "-Wno-*" statements so that these warnings dont end up as errors, allowing the compilation to proceed. 

#### Risks (if any) associated the changes in the commit

The change only affects the build if the compiler is clang, which was not supported up to this point anyway. For GCC the substitution should yield exactly the same result.

#### What has been tested and how, request additional testing if necessary

This was tested with current master XRT and clang-12 on Ubuntu 20.04.03 LTS.

keryell commented:
Great!
Looking at the changes, @stsoe it looks like that means someone should spend some time beyond this PR.
What about assigning a new employee on this as a nice way to discover our development tools and code architecture? :-)

stsoe commented:
ok to test - awesome thanks so much @HFTrader

gbuildx commented:
Build Passed!

HFTrader commented:
@stsoe No problem, always willing to help.
@keryell when I said "it's dangerous" I did not mean the warning, it's the code that's already there.

HFTrader commented:
I did not realize that by pushing more changes to my XRT fork it would merge into here. 
However, these changes are related to the original push. 
Initially I had added a section in CMakeLists.txt just to ignore the issues that Clang was finding. Now, I actually went in and addressed those issues. A couple of them deserve attention as they might break in production.

HFTrader commented:
The last standing compiler warning -Wno-overloaded-virtual  is a thorny one. There are several places in the code where virtual methods are overridden with methods with the same name but different arguments. This won't be easily solved without much refactoring.

stsoe reviewed (commented):
Wonderful.  Thank you for the effort @HFTrader.
One comment, but I don't really care if it is addressed.

I am curious about a case of `-Woverload-virtual` and whether it is a real problem that we should fix?  Can you point to a few places where this warning occurs?

gbuildx commented:
Build Passed!

HFTrader commented on src/runtime_src/xrt/device/device.h:
```diff
@@ -65,7 +65,7 @@ class device : public xrt_device
   {}
 
-  ~device()
+  virtual ~device()
```
xocl::image::get_buffer_object
```bash
XRT/src/runtime_src/xocl/core/memory.h:742:3: error: 'xocl::image::get_buffer_object' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  get_buffer_object(device* device);
  ^
XRT/src/runtime_src/xocl/core/memory.h:338:3: note: hidden overloaded virtual function 'xocl::memory::get_buffer_object' declared here: different number of parameters (2 vs 1)
  get_buffer_object(device* device, memidx_type subidx=-1);
  ^
```

xrt::run_impl::ps_arg_setter::set_arg_value
```bash
XRT/src/runtime_src/core/common/api/xrt_kernel.cpp:1744:5: error: 'xrt::run_impl::ps_arg_setter::set_arg_value' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
    set_arg_value(const argument& arg, const xrt::bo& bo) override
    ^
XRT/src/runtime_src/core/common/api/xrt_kernel.cpp:1688:5: note: hidden overloaded virtual function 'xrt::run_impl::hs_arg_setter::set_arg_value' declared here: type mismatch at 2nd parameter ('const arg_range<uint8_t> &' (aka 'const arg_range<unsigned char> &') vs 'const xrt::bo &')
    set_arg_value(const argument& arg, const arg_range<uint8_t>& value) override
    ^
```

xrt_core::device_linux::import_bo
```bash
XRT/src/runtime_src/core/pcie/linux/device_linux.h:67:3: error: 'xrt_core::device_linux::import_bo' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  import_bo(pid_t pid, xclBufferExportHandle ehdl) override;
  ^
XRT/src/runtime_src/core/common/ishim.h:302:3: note: hidden overloaded virtual function 'xrt_core::shim<xrt_core::device_pcie>::import_bo' declared here: different number of parameters (1 vs 2)
  import_bo(xclBufferExportHandle ehdl) override
  ^
```

xrt_core::swemu::system::program_plp
```bash
XRT/src/runtime_src/core/pcie/emulation/cpu_em/generic_pcie_hal2/system_swemu.h:42:3: error: 'xrt_core::swemu::system::program_plp' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  program_plp(const device* dev, const std::vector<char> &buffer) const;
  ^
XRT/src/runtime_src/core/common/system.h:127:3: note: hidden overloaded virtual function 'xrt_core::system::program_plp' declared here: different number of parameters (3 vs 2)
  program_plp(const device*, const std::vector<char>&, bool) const
  ^
```

xrt_core::hwemu::system::program_plp
```bash
XRT/src/runtime_src/core/pcie/emulation/hw_em/generic_pcie_hal2/system_hwemu.h:42:3: error: 'xrt_core::hwemu::system::program_plp' hides overloaded virtual function [-Werror,-Woverloaded-virtual]
  program_plp(const device* dev, const std::vector<char> &buffer) const;
  ^
XRT/src/runtime_src/core/common/system.h:127:3: note: hidden overloaded virtual function 'xrt_core::system::program_plp' declared here: different number of parameters (3 vs 2)
  program_plp(const device*, const std::vector<char>&, bool) const
  ^
```
